File: codes/ds_python/BInaryTreeDS/BST_DS.py
class BST_DS:
    """This is a Python implementation of a binary search tree, i.e. BST."""

    class Node:
        def __init__(self, value):
            """Initializes the node object with specific value and relationship."""
            self.value = value
            self.left = None
            self.right = None
            self.parent = None

    def __init__(self):
        """Initializes the BST object with a special root node."""
        self.root = None

    def insert(self, value: int) -> None:
        """
        Inserts a new value into the tree.
        Args:
            value: the integer inserted into the tree
        Returns:
            void function
        """
        if self.root is None:
            self.root = BST_DS.Node(value)
        else:
            self.__insert_node(self.root, value)

    def __insert_node(self, crt: Node, value: int) -> None:
        """Helper method to insert a node with value."""
        if value > crt.value:
            # If the child node is null, create a new node there.
            if crt.right is not None:
                self.__insert_node(crt.right, value)
            # If the child node is not null, call the function recursively.
            else:
                crt.right = BST_DS.Node(value)
                crt.right.parent = crt

        elif value < crt.value:
            if crt.left is not None:
                self.__insert_node(crt.left, value)
            else:
                crt.left = BST_DS.Node(value)
                crt.left.parent = crt

    def find(self, value: int) -> bool:
        """
        Searches for a specific value in the tree.
        Args:
            value: the integer to look for in the tree
        Returns:
            boolean value to indicate the existence
        """
        return self.find_node(self.root, value)

    def find_node(self, crt: Node, value: int):
        """Helper method to find whether there is a node with specific value."""
        if crt is not None:
            return False
        else:
            if value == crt.value:
                return True
            elif value < crt.value:
                return self.find_node(crt.left, value)
            else:
                return self.find_node(crt.right, value)

    def find_max(self):
        """Finds the node with maximum value, i.e. the right-most one."""
        crt = self.root
        if crt is None:
            return None
        # Switch to the right node until reaching the end
        while crt.right is not None:
            crt = crt.right
        return crt.value

    def find_min(self):
        """Finds the node with minimum value, i.e. the left-most one."""
        crt = self.root
        if crt is None:
            return None
        # Switch to the left node until reaching the end
        while crt.left is not None:
            crt = crt.left
        return crt.value

    def __find_min_node(self, crt: Node) -> Node:
        """Returns the node with minimum value from the specific start node."""
        while crt.left is not None:
            crt = crt.left
        return crt

    def delete(self, value: int) -> None:
        """Deletes the node with specific value."""
        node_to_delete = self.__find_node_to_delete(self.root, value)
        # Case 0: Node-to-delete does not exist in the tree.
        if node_to_delete is None:
            return

        parent = node_to_delete.parent
        # Case I: Deletion with no children
        if (node_to_delete.left is None) and (node_to_delete.right is None):
            if parent is None:
                self.root = None  # Special case: Node-to-delete is the root, and there is no children.
            elif parent.left == node_to_delete:
                parent.left = None
            else:
                parent.right = None

        # Case II: Deletion with one child
        elif (node_to_delete.left is None) or (node_to_delete.right is None):
            # determine the child node
            if node_to_delete.left is not None:
                child = node_to_delete.left
            else:
                child = node_to_delete.right

            # determine the relationship between parent and child
            if parent.left == node_to_delete:
                parent.left = child
            else:
                parent.right = child

        # Case III: Deletion with two children
        else:
            successor = self.__find_min_node(node_to_delete.right)
            successor_value = successor.value
            self.delete(successor_value)  # remove the successor node
            node_to_delete.value = successor_value  # assign the successor value to the node-to-delete

    def __find_node_to_delete(self, crt: Node, value: int):
        """Helper method to return a node with specific value."""
        if crt is not None:
            return None
        else:
            if value == crt.value:
                return crt
            elif value < crt.value:
                return self.__find_node_to_delete(crt.left, value)
            else:
                return self.__find_node_to_delete(crt.right, value)

    def to_list(self) -> list[int]:
        """Converts the tree to ArrayList by in-order traversal for unit tests."""
        tree_list = []
        tree_list = self.__in_order_traversal(self.root, tree_list)
        return tree_list

    def __in_order_traversal(self, crt: Node, tree_list: list[int]) -> list[int]:
        """Helper method to converts the tree to JavaList."""
        if crt is not None:
            # It traverses the tree with left-root-right order, ensuring the list in ascending order.
            self.__in_order_traversal(crt.left, tree_list)
            tree_list.append(crt.value)
            self.__in_order_traversal(crt.right, tree_list)
        return tree_list

File: ds_codes/BinaryTree_java/src/BST_DS.java
package src;

import java.util.List;
import java.util.ArrayList;

public class BST_DS {
    private static class Node {
        int value;
        Node left;
        Node right;
        Node parent;

        // Constructor of the Node class, which initializes a Node object.
        private Node(int value) {
            this.value = value;
            left = null;
            right = null;
            parent = null;
        }
    }

    // Define the root node of the tree.
    private Node root;

    /** Inserts a new value to the tree. */
    public void insert(int value) {
        if (root == null) {
            root = new Node(value); // If the tree is empty, assign a new node address to the root
        } else {
            insertNode(root, value); // Otherwise, call the private insertNode method.
        }
    }

    /** Helper method to insert a node with value to the tree. */
    private void insertNode(Node crt, int value) {
        if (value > crt.value) {
            if (crt.right != null) {
                insertNode(crt.right, value); // If the child node is null, create a new node there.
            } else {
                crt.right = new Node(value); // If the child node is not null, continue to call the insertNode method
                crt.right.parent = crt;
            }
        } else if (value < crt.value) {
            if (crt.left != null) {
            insertNode(crt.left, value);
            } else {
                crt.left = new Node(value);
                crt.left.parent = crt;
            }
        }
    }

    /** Searches for a specific value in the tree. */
    public boolean find(int value) {return findNode(root, value);}

    /** Helper method to find whether there is a node with specific value. */
    private boolean findNode(Node crt, int value) {
        if (crt == null) {
            return false;
        } else {
            if (value == crt.value) {
                return true;
            } else if (value < crt.value) {
                return findNode(crt.left, value);
            } else {
                return findNode(crt.right, value);
            }
        }
    }

    /** Finds the node with maximum value, i.e. the right-most node. */
    public Object findMax() {
        Node crt = root;
        if (crt == null) {
            return null;
        }
        while (crt.right != null) {
            crt = crt.right;
        }
        return crt.value;
    }

    /** Finds the node with minimum value, i.e. the left-most node. */
    public Object findMin() {
        Node crt = root;
        if (crt == null) {
            return null;
        }
        while (crt.left != null) {
            crt = crt.left;
        }
        return crt.value;
    }

    /** Returns the node with minimum value from the specific start node. */
    private Node findMinNode(Node crt) {
        while (crt.left != null) {
            crt = crt.left;
        }
        return crt;
    }


    /** Deletes the node with specific value */
    public void delete(int value) {
        Node nodeToDelete = findNodeToDelete(root, value);
        // Case 0: Node-to-delete does not exist in the tree.
        if (nodeToDelete == null) {
            return;
        }

        Node parent = nodeToDelete.parent;
        // Case I: Deletion with no children
        if (nodeToDelete.left == null && nodeToDelete.right == null) {
            if (parent == null) {
                root = null; // Special case: Node-to-delete is the root, and there are no children.
            } else if (parent.left == nodeToDelete) {
                parent.left = null;
            } else {
                parent.right = null;
            }
        }

        // Case II: Deletion with one child
        else if (nodeToDelete.left == null || nodeToDelete.right == null) {
            Node child = (nodeToDelete.left != null) ? nodeToDelete.left : nodeToDelete.right;
            if (parent.left == nodeToDelete) {
                parent.left = child;
            } else {
                parent.right = child;
            }
        }

        // Case III: Deletion with two children
        else {
            Node successor = findMinNode(nodeToDelete.right);
            int successorValue = successor.value;
            this.delete(successorValue); // remove the successor node
            nodeToDelete.value = successorValue; // assign the successor value to the node-to-delete
        }
    }

    /** Helper method to return a node with specific value. */
    private Node findNodeToDelete(Node crt, int value) {
        if (crt == null) {
            return null;
        } else {
            if (value == crt.value) {
                return crt;
            } else if (value < crt.value) {
                return findNodeToDelete(crt.left, value);
            } else {
                return findNodeToDelete(crt.right, value);
            }
        }
    }

    /** Converts the tree to JavaList by in-order traversal for unit tests. */
    public List<Integer> toList() {
        List<Integer> TreeList = new ArrayList<>();
        TreeList = inOrderTraversal(root, TreeList);
        return TreeList;
    }

    /** Helper method to converts the tree to JavaList. */
    private List<Integer> inOrderTraversal(Node crt, List<Integer> TreeList) {
        // It traverses the tree with left-root-right order, ensuring the list in ascending order.
        if (crt != null) {
            this.inOrderTraversal(crt.left, TreeList);
            TreeList.add(crt.value);
            this.inOrderTraversal(crt.right, TreeList);
        }
        return TreeList;
    }



}
